File: leetcode/dynamic_programming/322_coinChange/322_coinChange.js
/**
 * Created by betty on 9/5/20.
 */
"use strict";
//@ts-ignore
let coinChange = (coins, amount) => {
  let mat = new Array(amount + 1)
  mat[0] = 0
  for (let i = 1; i < amount + 1; i++) {
    mat[i] = amount + 1
  }
  for (let a = 1; a <= amount; a++) {
    for (let j = 0; j < coins.length; j++) {
      let alt = amount + 1;
      if (a >= coins[j]) {
        alt = mat[a - coins[j]] + 1
      }
      mat[a] = Math.min(mat[a], alt);
    }
  }
  return mat[amount] < amount + 1 ? mat[amount] : -1
};

File: leetcode/dynamic_programming/322_coinChange/322_coinChange.go
package _22_coinChange

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func CoinChange(coins []int, amount int) int {
	arr := make([]int, amount+1)
	for i := 1; i < amount+1; i++ {
		arr[i] = amount + 1
	}
	for i := 1; i < amount+1; i++ {
		for j := 0; j < len(coins); j++ {
			if i >= coins[j] {
				arr[i] = min(arr[i], arr[i-coins[j]]+1)
			}
		}
	}
	if arr[amount] <= amount {
		return arr[amount]
	} else {
		return -1
	}
}
